src/main/scala/org/deepdive/ddlog/DeepDiveLogCompiler.scala

```diff
@@ -767,193 +521,152 @@ ${if (createTable) {
+        case FactorFunction.Imply()  => "Imply"
+        case FactorFunction.And()    => "And"
+        case FactorFunction.Or()     => "Or"
+        case FactorFunction.Equal()  => "Equal"
+        case FactorFunction.Linear() => "Linear"
+        case FactorFunction.Ratio()  => "Ratio"
+        case FactorFunction.IsTrue() =>
+          schemaDeclarationByRelationName get(stmt.head.terms(0).name) map(_.variableType) match {
+            case Some(Some(MultinomialType(_))) => "Multinomial"
+            case _ => "Imply" // TODO fix to IsTrue
+          }
+        case FactorFunction.Multinomial() => "Multinomial"
+      }
+      s"""${function}(${funcBody.mkString(", ")})"""
     }
-    List("""
-  deepdive.db.default {
-    driver: "org.postgresql.Driver"
-    url: "jdbc:postgresql://"${PGHOST}":"${PGPORT}"/"${DBNAME}
-    user: ${PGUSER}
-    password: ${PGPASSWORD}
-    dbname: ${DBNAME}
-    host: ${PGHOST}
-    port: ${PGPORT}
-    incremental_mode: """ + s"""${mode}
+
+    // weight
+    val weight = {
+      // note error cases should be handled in semantic checker
+      // FIXME check if all weights.variables are ConstExpr to determine whether it's a fixed or unknown weight
+      stmt.weights.variables(0) match {
+        case IntConst(value) => value.toString
+        case DoubleConst(value) => value.toString
+        case StringConst(value) => "?"
+        case _ => {
+          val weightVars = stmt.weights.variables.zipWithIndex.flatMap {
+            case(s: Expr, i) => Some(s"dd_weight_column_${i}")
+          } mkString(", ")
+          s"?(${weightVars})"
+        }
+      }
     }
-    """)
+
+    List(s"deepdive.inference.factors.${resolveInferenceBlockName(stmt)}" -> Map(
+        "input_query" -> QuotedString(inputQueries.mkString(" UNION ALL ")),
+        "function" -> QuotedString(func),
+        "weight" -> QuotedString(weight),
+        "input_relations" -> {
+          val relationsInHead = stmt.head.terms map (_.name)
+          val relationsInBody = relationNamesUsedByStatement getOrElse(stmt, List.empty)
+          (relationsInHead ++ relationsInBody) distinct
+        }
+    ))
   }
 
-  def compileVariableKey(ss: CompilationState): CompiledBlocks = {
-    var keys = new ListBuffer[String]()
-    for (stmt <- (ss.schemaDeclarationGroupByHead map (_._2)).flatten) {
-      if (stmt.isQuery) keys += s"""${stmt.a.name} : [${stmt.a.terms.mkString(", ")}]"""
-    }
-    ss.mode match {
-      case INCREMENTAL => List(s"""
-      deepdive.schema.keys {
-        ${keys.mkString("\n        ")}
-      }""")
-      case _ => List("")
-    }
+  def compile(stmt: Statement): CompiledBlocks = stmt match {
+    case s: SchemaDeclaration => compile(s)
+    case s: FunctionDeclaration => compile(s)
+    case s: ExtractionRule => compile(s)
+    case s: FunctionCallRule => compile(s)
+    case s: InferenceRule => compile(s)
+    case _ => sys.error(s"Compiler does not recognize statement: ${stmt}")
```

Comment:
This is a common misbelief.  Due to JVM's design the function call target is determined at compile time.  Dynamic dispatch works if `compile()` was a method on the Statement class following the visitor pattern, but here we don't have such setup.  We could try that at some point decorating the parse tree classes since pretty printer and semantic checker all have these variants.  Not sure if Scala provides any help here.